### PR TITLE
use fflDB to store block

### PR DESCRIPTION
**Pull request by RainFallsSilent**

### CODE REVIEW

File: blockchain/ledgerstore.go [hunk 1]
```diff
@@ -6,14 +6,20 @@
 package blockchain
```
Comment (peterwillcn) [inline blockchain/ledgerstore.go]: Add license

File: blockchain/upgrade.go [hunk 2]
```diff
@@ -0,0 +1,223 @@
+package blockchain
```
Comment (peterwillcn) [inline blockchain/upgrade.go]: Add license

File: blockchain/ledgerstore.go (hunk 1 above)
Comment (RainFallsSilent) [reply blockchain/ledgerstore.go]: exist license before

File: blockchain/upgrade.go (hunk 2 above)
Comment (RainFallsSilent) [reply blockchain/upgrade.go]: added

File: blockchain/upgrade.go [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+//
+
+package blockchain
+
+import (
+	"bytes"
+	"container/list"
+	"errors"
+	"fmt"
+	"github.com/elastos/Elastos.ELA/core/types"
```
Comment (Houshoupei84) [inline blockchain/upgrade.go]: import format

File: cmd/script/api/api.go [hunk 4]
```diff
@@ -172,29 +172,35 @@ func initLedger(L *lua.LState) int {
...
 
 	var interrupt = signal.NewInterrupt()
-	chain, err := blockchain.New(chainStore, chainParams,
+	chain, initialized, err := blockchain.New(chainStore, chainParams,
 		state.NewState(chainParams, arbiters.GetArbitrators, nil), nil)
 	if err != nil {
 		fmt.Printf("Init block chain error: %s \n", err.Error())
 	}
+	if !initialized {
+		chain.InitFFLDBFromChainStore(interrupt.C, nil, nil)
```
Comment (Houshoupei84) [inline cmd/script/api/api.go]: do we need consider InitFFLDBFromChainStore return error ?

File: cmd/script/api/api.go [hunk 5]
```diff
@@ -9,6 +9,7 @@ import (
 	"bytes"
 	"encoding/hex"
 	"fmt"
+	"github.com/elastos/Elastos.ELA/core/types"
```
Comment (Houshoupei84) [inline cmd/script/api/api.go]: import format

File: blockchain/ledgerstore.go [hunk 6]
```diff
@@ -6,14 +6,20 @@
...
 
 import (
+	"github.com/elastos/Elastos.ELA/database"
```
Comment (Houshoupei84) [inline blockchain/ledgerstore.go]: import format

File: blockchain/upgrade.go (hunk 3 above)
Comment (RainFallsSilent) [reply blockchain/upgrade.go]: fixed

File: cmd/script/api/api.go (hunk 4 above)
Comment (RainFallsSilent) [reply cmd/script/api/api.go]: fixed

File: cmd/script/api/api.go (hunk 5 above)
Comment (RainFallsSilent) [reply cmd/script/api/api.go]: fixed

File: blockchain/ledgerstore.go (hunk 6 above)
Comment (RainFallsSilent) [reply blockchain/ledgerstore.go]: fixed

File: blockchain/blockchain.go [hunk 7]
```diff
@@ -298,8 +380,11 @@ func (b *BlockChain) ContainsTransaction(hash Uint256) bool {
 	return true
 }
 
-func (b *BlockChain) CurrentBlockHash() Uint256 {
-	return b.db.GetCurrentBlockHash()
+func (b *BlockChain) GetCurrentBlockHash() Uint256 {
+	b.IndexLock.RLock()
+	defer b.IndexLock.RUnlock()
+
+	return *b.Nodes[len(b.Nodes)].Hash
```
Comment (Houshoupei84) [inline blockchain/blockchain.go]: minus one

Comment (RainFallsSilent) [reply blockchain/blockchain.go]: fixed

File: glide.yaml [hunk 8]
```diff
@@ -25,5 +25,6 @@ import:
 - package: github.com/yuin/gopher-lua
 - package: gopkg.in/cheggaaa/pb.v1
 - package: gopkg.in/yaml.v2
+- package: github.com/btcsuite/btcd/wire
```
Comment (yz89) [inline glide.yaml]: Is it too big to reference the whole package? @raindust @RainFallsSilent

Comment (raindust) [reply glide.yaml]: Yes but we can improve it later when meta-data about transactions and UTXOs are well defined

**Pull request merged**